rfcs/010-gateway-refactor/010-gateway-refactor.md

```diff
@@ -0,0 +1,211 @@
+- Start Date: 09-12-2022
+- Status: Proposed
+
+# Gateway refactor
+
+## Summary
+
+This proposes a full rewrite of the current gateway service. In addition, it also proposes
+changes to the way that access control to Jupyter servers is handled. Currently, each Jupyter
+server establishes a session with the user. But going forward each user will have a session
+with the gateway which will be used to control access into each Jupyter server. 
+The main motivation for changing the access control for Jupyter servers is the desire to run
+each Jupyter server in its own subdomain. The problem with unique subdomains is that each Jupyter
+server's subdomain would have to be registered in Keycloak so that the Oauth2-proxy in each 
+Jupyter server can function properly. But if the Oauth session is handled by the gateway we can
+have a single callback address in Keycloak as we currently do.
+
+## Motivation
+
+Reasons for rewriting the gateway:
+- the current code base is hard to maintain
+- one service could replace both traefik and the gateway
+- problems with managing sessions across different services/clients
+- using golang should result in more efficient use of resources and improved performance
+
+Reasons for centralized session authentication:
+- running sessions in subdomains which would allow for SSH access
+
+## Design Detail
+
+### Gateway
+
+The gateway has to have sessions with different clients. Some of these are:
+- User servers (i.e. the sidecar container needs be able to swap some kind of ID
+or access token for a Gitlab token)
+- Actual users logging in through the UI
+- Actual users logging in through the CLI
+
+Using JWT access tokens for this purpose is not feasible. This is because access
+tokens from Keycloak expire usually within an hour or several hours. This means
+that anyone who has to remain authenticated for a long time with the gateway
+needs to also get a refresh token and worry about using this to have valid access
+tokens. Accepting an expired access token is not secure.
+
+The gateway should instead have server-side sessions with every "client". For each such
+session the gateway will issue a random, hard to guess ID that will be stored either
+in a cookie or simply as a secret. This is exactly how sessions are handled by the
+ui-server. The cookie or ID is presented with every request to the gateway. The gateway
+looks this ID up in redis and if there is a match then any tokens/credentials stored
```

Comment:
Would it make sense to use something like HashiCorp Vault instead of Redis?

Reply:
We may end up using Vault for storing static secrets from users such as S3 keys, or database credentials, etc. But I think using Vault for this is overkill. Storing secrets in redis in the backend for a user session is standard IMO. Also not sure if Vault will be as performant as Redis when it comes to having many requests that continuously and periodically check all tokens and refresh them if needed. This is probably not that big of a deal for secrets that are static, but it may cause problems for oauth access/refresh tokens.

Reply:
@wesjdj have you seen or used of Vault to store oauth2 tokens? I will do some googling to see if this is feasible.

Reply:
@aledegano might have some more insight on this but I think Vault is quite performant and can store whatever is asked of it

Reply:
So on this topic I stumbled upon this issue on vault: https://github.com/hashicorp/vault/issues/96 which then lead me to this vault plugin: https://github.com/puppetlabs/vault-plugin-secrets-oauthapp. This makes things more interesting because with the plugin vault would automatically refresh the oauth tokens we store. But without the plugin we would have to do this ourselves.

But I know very little about vault and have never used in production. The perspective on this from someone with more experience would be definitely appreciated.

Reply:
oh that's a very interesting option! Some vault ninja should try it against the dev gitlab...

Reply:
I can try that when I'm back from x-mas holidays.